Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp

```diff
@@ -0,0 +1,152 @@
+        g_main_loop_quit(test->m_mainLoop);
+    }
+
+    void waitUntilIsImmersiveModeEnabledChanged()
+    {
+        g_signal_connect(m_webView.get(), "notify::is-immersive-mode-enabled", G_CALLBACK(isImmersiveModeEnabledChanged), this);
+        g_main_loop_run(m_mainLoop);
+    }
+
+    bool isImmersiveModeEnabled() const
+    {
+        return webkit_web_view_is_immersive_mode_enabled(m_webView.get());
+    }
+
+    void leaveImmersiveMode()
+    {
+        webkit_web_view_leave_immersive_mode(m_webView.get());
+    }
+
+    struct NetworkPolicyGuard {
+        NetworkPolicyGuard(ImmersiveModeTest* test, WebKitTLSErrorsPolicy policy)
+            : m_test(test)
+        {
+            // Ignore server certificate errors.
+#if ENABLE(2022_GLIB_API)
+            m_originalPolicy = webkit_network_session_get_tls_errors_policy(m_test->m_networkSession.get());
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), policy);
+#else
+            m_manager = webkit_web_context_get_website_data_manager(m_test->m_webContext.get());
+            m_originalPolicy = webkit_website_data_manager_get_tls_errors_policy(m_manager);
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, policy);
+#endif
+        }
+        ~NetworkPolicyGuard()
+        {
+#if ENABLE(2022_GLIB_API)
+            webkit_network_session_set_tls_errors_policy(m_test->m_networkSession.get(), m_originalPolicy);
+#else
+            webkit_website_data_manager_set_tls_errors_policy(m_manager, m_originalPolicy);
+#endif
+        }
+        ImmersiveModeTest* m_test;
+        WebKitTLSErrorsPolicy m_originalPolicy;
+#if !ENABLE(2022_GLIB_API)
+        WebKitWebsiteDataManager* m_manager;
+#endif
+    };
+};
+
+#if USE(SOUP2)
+static void serverCallback(SoupServer* server, SoupMessage* message, const char* path, GHashTable*, SoupClientContext*, gpointer)
+#else
+static void serverCallback(SoupServer* server, SoupServerMessage* message, const char* path, GHashTable*, gpointer)
+#endif
+{
+    if (soup_server_message_get_method(message) != SOUP_METHOD_GET) {
+        soup_server_message_set_status(message, SOUP_STATUS_NOT_IMPLEMENTED, nullptr);
+        return;
+    }
+
+    auto* responseBody = soup_server_message_get_response_body(message);
+
+    if (g_str_equal(path, "/xr-session/")) {
+        soup_server_message_set_status(message, SOUP_STATUS_OK, nullptr);
+        soup_message_body_append(responseBody, SOUP_MEMORY_STATIC, indexHTML, strlen(indexHTML));
+        soup_message_body_complete(responseBody);
+    } else {
+        g_assert(false);
+        soup_server_message_set_status(message, SOUP_STATUS_NOT_FOUND, nullptr);
+    }
+}
+
+static void testWebKitImmersiveModeLeaveImmersiveMode(ImmersiveModeTest* test, gconstpointer)
+{
+    auto httpsServer = makeUnique<WebKitTestServer>(WebKitTestServer::ServerHTTPS);
+    httpsServer->run(serverCallback);
```

Comment:
This is ok because we only have one test, but we normally make the server global, created in beforeAll and deleted in afterAll